### PR TITLE
drivers: spi: add PolarFire SoC SPI driver

### PR DESCRIPTION
**Pull request by con-pax**
This patch set adds support for the SPI driver on Microchip's PolarFire SoC Icicle Kit.

### CODE REVIEW

File: drivers/spi/spi_mchp_mss.c
Comment (tbursztyka) [inline drivers/spi/spi_mchp_mss.c:300]: Is there a reason why you specifically handle each interrupt the same way but separately? 

Couldn't it be handled this way:
```
mss_spi_write(cfg, MSS_SPI_REG_INT_CLEAR, intfield);
spi_context_complete(ctx, dev, 0);
```

And more importantly: where do you actually act on the relevant interrupt?
Why the logic of mss_spi_readwr_fifo() is not being called in ISR mode?

If this controller is able to run in interrupt mode, logically you would configure/enable/unmask INTs and... in the interrupt handler here, you would move bytes in/out depending on the interrupt type.

File: drivers/spi/spi_mchp_mss.c
Comment (tbursztyka) [inline drivers/spi/spi_mchp_mss.c:366]: You don't seem to return -ENOTSUP if the frame size is != 8 bits. 
See macro SPI_WORD_SIZE_GET() from include/zephyr/drivers/spi.h

File: drivers/spi/spi_mchp_mss.c
Comment (tbursztyka) [inline drivers/spi/spi_mchp_mss.c:399]: Move this into configure function. The point of this is also to avoid reconfiguring the controller if ctx->config is the same again as spi_cfg on a new transceive call. Unless your controller needs to be fully reconfigured every time.

**Review by tbursztyka — DISMISSED** [review]
Perhaps this is controller specific, but the logic seems a bit off.

File: drivers/spi/spi_mchp_mss.c
Comment (nagasureshkumar) [reply drivers/spi/spi_mchp_mss.c:300]: > Is there a reason why you specifically handle each interrupt the same way but separately?
> 
> Couldn't it be handled this way:
Thanks Tomasz for your time to review this PR.
Yes the below way is the cleanest way. I will update it.
> 
> ```
> mss_spi_write(cfg, MSS_SPI_REG_INT_CLEAR, intfield);
> spi_context_complete(ctx, dev, 0);
> ```
> 
> And more importantly: where do you actually act on the relevant interrupt? Why the logic of mss_spi_readwr_fifo() is not being called in ISR mode?
> 
> If this controller is able to run in interrupt mode, logically you would configure/enable/unmask INTs and... in the interrupt handler here, you would move bytes in/out depending on the interrupt type.
I agree with your explanation about generically handling tx and rx from interrupt handler.
But we configure the controller in such a way that, the TX DONE will be set once the configured number of transmit bytes are sent and the RXRDY will be set once all the configured number of rx bytes are read from the RXFIFO(from the mss_spi_readwr_fifo() functions). hence in the interrupt handler we just acknowledge the bits as the entire data transmission and reception is handled inside mss_spi_readwr_fifo().
But agree that, in this case the interrupt handler is very small as mentioned by you like mss_spi_write(cfg, MSS_SPI_REG_INT_CLEAR, intfield);

File: drivers/spi/spi_mchp_mss.c
Comment (nagasureshkumar) [reply drivers/spi/spi_mchp_mss.c:366]: > You don't seem to return -ENOTSUP if the frame size is != 8 bits. See macro SPI_WORD_SIZE_GET() from include/zephyr/drivers/spi.h

Yes. I will add that
thanks for pointing the helper function.

File: drivers/spi/spi_mchp_mss.c
Comment (nagasureshkumar) [reply drivers/spi/spi_mchp_mss.c:370]: > Move this at the top, it's better to fail the earliest as possible

Ok. I will update it.

File: drivers/spi/spi_mchp_mss.c
Comment (nagasureshkumar) [reply drivers/spi/spi_mchp_mss.c:375]: > It would be better to use a macro for this value '0xC'.
> 
> btw is this the enablement of the controller? If so, move it out of configure, once the buffers are setup.

Sure. I will update it.

File: drivers/spi/spi_mchp_mss.c
Comment (nagasureshkumar) [reply drivers/spi/spi_mchp_mss.c:399]: > Move this into configure function. The point of this is also to avoid reconfiguring the controller if ctx->config is the same again as spi_cfg on a new transceive call. Unless your controller needs to be fully reconfigured every time.

Ok. Will update it.

File: drivers/spi/spi_mchp_mss.c
Comment (nagasureshkumar) [reply drivers/spi/spi_mchp_mss.c:452]: > I don't know anything about this controller, but you seem to disable/enable the controller without doing anything to it. xfer->control is not written anywhere.
> 
> Btw, are you sure that the controller must stay enabled while not transferring? Should it be enabled before configuring, and disabled in case of error or at the end of the transaction?

oh, understood. It should be enabled at the start of transfer and disable after completion of transfer.
Let me recheck the configuration, while developing the driver I have seen that with out disabling the core, I can't be able to set any of the bits in the controller register. But as mentioned by you, let me re frame the controller enable/disable task.

File: drivers/spi/spi_mchp_mss.c
Comment (tbursztyka) [reply drivers/spi/spi_mchp_mss.c]: so you loose the interest of having the controller running on interrupt mode, no? It's like polling mode

File: drivers/spi/spi_mchp_mss.c
Comment (tbursztyka) [reply drivers/spi/spi_mchp_mss.c]: I missed that, but beware there is a difference with enabled to transfer and enabled about power.  If that's about power, just leave it out.

**Comment by kgugala** [comment]
the PR looks good to me, however I'm not the SPI driver area maintainer. I suppose @tbursztyka should have the final word here.

@con-pax can you rebase the PR - there are some conflicts

**Comment by con-pax** [comment]
> @con-pax can you rebase the PR - there are some conflicts

Hey @kgugala, 
Thanks a million. Yep I will rebase today, test and push.

File: drivers/spi/spi_mchp_mss.c
Comment (nagasureshkumar) [reply drivers/spi/spi_mchp_mss.c]: Hi @tbursztyka,

Sorry for the late reply. its about enabled to transfer. as you mentioned previously, I reframed the controller enable/disable task. In the initial version of the driver the controller enable/disable is taken place at many places. but in this version, as you said it is enabled before configuration and disabled at the end of transfer.

**Review by kgugala — DISMISSED** [review]
LGTM

**Review by cfriedt — DISMISSED** [review]
LGTM

File: drivers/spi/spi_mchp_mss.c
Comment (con-pax) [reply drivers/spi/spi_mchp_mss.c]: Hi @tbursztyka, could I ask you to confirm your changes have been addressed so that we can get this PR merged?
Thanks a million!

**Comment by con-pax** [comment]
@kgugala @cfriedt @tbursztyka 
It seems as though all comments have been resolved on this PR. I am going to rebase test and push. When done, If it's not too much trouble could I ask for your ACK again? We are working on some FPGA services that require SPI, so it would be really cool to get this in soon. 
Cheers!

**Comment by teburd** [comment]
Does the polarfire spi ip have a loopback mode? If so have you ran the spi tests for it?

**Comment by nagasureshkumar** [comment]
> Does the polarfire spi ip have a loopback mode? If so have you ran the spi tests for it?

Hi @teburd,

No, it does not have loopback mode.

**Comment by nagasureshkumar** [comment]
> Perhaps this is controller specific, but the logic seems a bit off.

As per the @tbursztyka review, all the comments are resolved and rebased the code.

**Comment by nagasureshkumar** [comment]
Hi @kgugala, @cfriedt @tbursztyka 
It seems @con-pax already rebased and the comments are resolved, Could you please ACK again if there are no other comments?
Thanks,
Naga Sureshkumar Relli.

**Review by cfriedt — APPROVED** [review]
Refresh+1

**Review by fkokosinski — APPROVED** [review]

**Comment by con-pax** [comment]
Hi @cfriedt @fkokosinski 
happy New Years to you! at the risk of being an absolute nuisance, it looks like I have to rebase again and push. If I can (hopefully for the last time) ask you to ACK again, it would be greatly appreciated. I don't really know why it didn't get merged the last time it had approvals.. it had the required approvals for a period of time. Maybe its because some of the conversations are not marked as resolved? anyway, I will rebase now, push and mark the conversations (that have actually been resolved) as resolved.

Cheers!

**Review by cfriedt — DISMISSED** [review]
Still LGTM 👍

**Review by fkokosinski — DISMISSED** [review]

**Comment by con-pax** [comment]
rebased to (hopefully) clear failing non-related CI

**Review by cfriedt — APPROVED** [review]

**Review by kgugala — APPROVED** [review]

**Comment by con-pax** [comment]
@tbursztyka Hi there,
Would it be possible for you to mark the changes that you requested as resolved to allow this PR to go through? 
Cheers!

**Comment by con-pax** [comment]
Hi @carlescufi,

I'm not sure what do on this PR. We would really appreciate this PR being merged. We have the required amount of approvals and implemented fixes requested. I have asked @tbursztyka to resolve the changes made, but haven't heard anything back.
Any help greatly appreciated :)
Cheers!

**Pull request merged**
